### PR TITLE
findAllReferences/rename: Search in all open projects

### PR DESCRIPTION
**Pull request by andy-ms**
Previously if a project A depended on a project B, we would find references in both only if the reference request started at A. Now the request can start in B, and we will find references in A as long project A has been opened.

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -738,6 +738,10 @@ namespace ts {
             getDiagnosticsProducingTypeChecker,
             getCommonSourceDirectory,
             emit,
+            getDeclarationEmitPath: fileName => {
+                const file = getSourceFile(fileName);
```
Comment (mhegazy) [inline src/compiler/program.ts]: why does this need to be on the program? `getDeclarationEmitOutputFilePath` is already available?

Comment (andy-ms) [reply src/compiler/program.ts]: `getEmitHost()` isn't though. That function appears to need `host.getCompilerOptions`, `host.getCurrentDirectory`, `host.getCommonSourceDirectory`, and `host.getCanonicalFileName`. `getCurrentDirectory` and `getCanonicalFileName` can be answered without a `Program`, but I don't think the other two can.

File: src/server/session.ts [hunk 2]
```diff
@@ -298,40 +298,55 @@ namespace ts.server {
...
     ): T[] {
         const outputs: T[] = [];
-        combineProjectOutputWorker(projects, <sourcemaps.SourceMappableLocation><any>undefined, projectService, ({ project }, tryAddToTodo) => {
-            for (const output of action(project)) {
-                if (!contains(outputs, output, resultsEqual) && !tryAddToTodo(project, getLocation(output))) {
-                    outputs.push(output);
+        combineProjectOutputWorker(
+            projects,
+            defaultProject,
+            { fileName: "<arbitrary>", position: 0 }, // Location is ignored by callback
```
Comment (sheetalkamat) [inline src/server/session.ts]: Why cant you pass null instead ? If location is ignored in ```combineProjectOutputWorker```  Dont think you should have need to create default value just for passing sake.

File: src/compiler/program.ts (hunk 1 above)
Comment (sheetalkamat) [reply src/compiler/program.ts]: getCurrentDirectory and getCanonicalFileName should be there on project and can be answered from there.

Comment (andy-ms) [reply src/compiler/program.ts]: That still leaves `getCompilerOptions` and `getCommonSourceDirectory` unfortunately.

File: src/server/session.ts [hunk 3]
```diff
@@ -345,89 +360,102 @@ namespace ts.server {
...
+            toDo = callbackProjectAndLocation(Debug.assertDefined(toDo.pop()), projectService, toDo, seenProjects, cb);
         }
     }
 
-    function callbackProjectAndLocation(
-        originalProjects: Projects, // For lazily populating seenProjects
-        projectAndLocation: ProjectAndLocation,
+    function getDefinitionInProject(definition: sourcemaps.SourceMappableLocation | undefined, definingProject: Project, project: Project): sourcemaps.SourceMappableLocation | undefined {
+        if (!definition || project.containsFile(toNormalizedPath(definition.fileName))) return definition;
+        const mappedDefinition = definingProject.getLanguageService().getSourceMapper().tryGetGeneratedLocation(definition);
```
Comment (sheetalkamat) [inline src/server/session.ts]: You also want to check if definingProject.languageServiceEnabled

File: src/compiler/program.ts (hunk 1 above)
Comment (sheetalkamat) [reply src/compiler/program.ts]: getCompilerOptions  is also present on project and getCommonSourceDirectory can be added as a private method to the project that redirects to the program ?

File: src/server/session.ts (hunk 3 above)
Comment (andy-ms) [reply src/server/session.ts]: That comes from `this.getDefaultProject`, shouldn't it always be enabled?

**Comment by andy-ms** [comment]
@sheetalkamat Please re-review

**Comment by mhegazy** [comment]
@andy-ms please port to release-3.0 as well

**Pull request merged**